urld73c47f3e5_lindenb_dataindexer_tgz-pJ8-test_github_lindenb_dataindexer_SimpleIntegerJ8/src/com/github/lindenb/dataindexer/AbstractDataIndexer.java

```diff
@@ -0,0 +1,64 @@
+package com.github.lindenb.dataindexer;
+
+import java.io.Closeable;
+import java.io.DataOutputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.util.logging.Logger;
+
+/**
+ * base class for both primary and secondary database writer
+ * @author lindenb
+ *
+ * @param <T> the class to be stored
+ * @param <CONFIG> the config type
+ */
+public abstract class AbstractDataIndexer<T,CONFIG extends AbstractConfig<T>>
+	implements Closeable
+	{
+	protected static final Logger LOG=Logger.getLogger("com.github.lindenb.dataindexer");
+
+	/** number of items stored */
+	protected long numberOfItems=0L;
+	/** data binding */
+	private TupleBinding<T> dataBinding=null;
+	
+	private CONFIG config;
+	protected AbstractDataIndexer(CONFIG cfg)
+		{
+		this.dataBinding=cfg.getDataBinding();
+		this.config=cfg;
+		
+		}
+	
+	protected void checkConfig() throws IOException
+		{
+		if(getConfig()==null) throw new IllegalStateException("config is null");
+		getConfig().validateForReading();
+		}
+	
+	/** close the underlying stream */
+	public abstract void close() throws IOException;
+	
+	/** returns the associated config */
+	public final CONFIG getConfig()
+		{
+		return this.config;
+		}
+	
+	/** get tuple binding for the type T */
+	protected final TupleBinding<T> getDataBinding()
+		{
+		return dataBinding;
+		}
+	
+	/** write the summary file */
+	protected void writeSummary()
+		throws IOException
+		{
+		DataOutputStream daos=new DataOutputStream(new FileOutputStream(getConfig().getSummaryFile()));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *daos*. The resource is referenced at lines 59, 60, 61. The resource is closed at line 62. There are other execution paths that do not contain closure statements, for example, when DataOutputStream.writeLong throws an exception. Ensure *daos* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).